### ISSUE TITLE
Unable to connect qz companion installed on NordicTrack Commercial S10i Studio Cycle

### ISSUE THREAD

**Issue by glotano** [cagnulein/qdomyos-zwift#1743]
**Describe the bug**
Unable to connect qz companion installed on NordicTrack Commercial S10i Studio Cycle.
I tried setting qz companion on s22i, studio bike pro 22, other, configuring the app on the phone accordingly. I also tried simulating treadmill 2950 and 1750 but the service installed on the bike does not communicate with the app on the phone.
As additional information, I add that I installed qz companion on the bike by installing the apk since the pc could not communicate with the bike OS via adb. It was finding port 5555 closed.
I attach the "qz fitness" logs extracted from the phone and a print screen showing that the service on the bike starts regularly

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):** _Version of Android installed on the Bike_ 
 - OS: Android
 - Version 5.1.1

**Smartphone (please complete the following information):**
 - Device: LG Q7 Thinq
 - OS: Android
 - Version 10

**Append a debug log**

  
[debug-Mon_Oct_30_11_57_53_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13204617/debug-Mon_Oct_30_11_57_53_2023.log)
![20231030_113329_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/6baf2a42-091f-4b64-a0de-307e4b21f851)

**Additional context**
Both during the tests I did yesterday and after resetting the bike again and reinstalling qz companion, I noticed that Ifit crashes from time to time (the application is unresponsive).

**Comment by cagnulein** [comment]
Hi @glotano quickly checked the log: there is a major issue on it. You don't have to put the ip in the TDF4 ip, but in the TDF companion ip instead. So blank the tdf4 one and press ok and set it to the companion and press ok and restart qz.
Send me a fresh log then
Thanks!

**Comment by glotano** [comment]
HI @cagnulein, I followed your suggestion, changing the settings. I attach a photo of them and the updated log.
Thank you!
![Screenshot_20231030-144940_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/630d2f3b-28d9-44fd-9f9b-b8eeac408b95)
[debug-Mon_Oct_30_14_44_23_2023.log](https://github.com/cagnulein/qdomyos-zwift/files/13205877/debug-Mon_Oct_30_14_44_23_2023.log)

**Comment by cagnulein** [comment]
ok perfect @glotano 
now I see that the companion is sending messages but they are empty because it can't read any log from the ifit tablet.
It could be possible because of 3 resons:
- one thing that I didn't say to you is that QZ must be running while you're in the manual window in ifit. You said that ifit is crashing, so you can't stay at all in the manual workout window?
- the companion can't read because it doesn't have the grep command. try to set the c1750 device in the companion
- ifit removed the log capabilities, I have already provided a workaround for this following https://github.com/cagnulein/qdomyos-zwift/issues/1591#issuecomment-1664105559
let me know

**Comment by cagnulein** [comment]
> I have to leave all on qz app unchanged (studio bike)?

yes on QZ, the only setting that matter is the tdf companion ip setting in that menu.



> I will also follow the third point.
> I will let you know as soon I will get the chance.

thanks!

**Comment by glotano** [comment]
Thanks to you @cagnulein! Your support is great.

**Comment by glotano** [comment]
Hi @cagnulein, I tried to run the command-line to get the permission, but the adb connection fails everytime, as when I tried to install the companion via adb, so I tried to install a tool on the bike's console to work locally but it fails anyway. It seems to me that the o.s. isn't able to start the service for debugging.
I attached a video to show you what happens.
Thank you!
Regards 
Giovanni

**Comment by cagnulein** [comment]
@glotano as far as I know Ifit closes adb on the latest update so try to reset the bike console and don't update ifit and enable adb. be careful: adb is resetting at every reboot but we just need to enable one time to set that permission

Let me know

Another similar case here https://github.com/cagnulein/qdomyos-zwift/issues/1706#issuecomment-1784188720

**Comment by glotano** [comment]
Hi @cagnulein , I tried to reset the bike's console more than one time today.
I installed a terminal and I realized that once enable the debugging mode and the usb debugging the bike's console doesn't start any daemon, so I started it manually using the command adb server-start.
I don't understand why the adb debugging bind the service on localhost on port 5038, so on the wrong ip and on the wrong port. For that reason the pc isn't able to establish a connection with it. I tried also to switch the binding on the port 5555 setting ANDROID_ADB_SERVER_PORT=5555, killing the daemon and restarting it, but I still found it binding on localhost port 5038.
I thought I could work locally at that point, to assign the permission, but I received an error.
Please take a look at the pics attached and let me know if you have some suggestion.
Thanks
Regards
Giovanni
![20231031_180602_HDR_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/87ed2958-ea5d-4d01-bc07-b72074d6c756)
![20231031_181415_HDR_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/79ff92eb-89c1-44c7-ada3-b820bec50a52)
![20231031_183206_HDR_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/995fa33c-4331-4277-8389-3a37b742d259)
![20231031_181415_HDR~2_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/a7da2ee9-935d-409a-a8d1-d3d2ae3973f2)

**Comment by cagnulein** [comment]
thanks i will check it asap

Il giorno mar 31 ott 2023 alle 20:59 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein, please Enjoy the evening and don't worry, it's not urgent.
;)

**Comment by glotano** [comment]
Hi @cagnulein, I tried to run logcat -f | grep Changed but I get the following error back:
"couldn't open output file: Read-Only file system".
I tried to run only logcat -f during the manual workout and I took a picture, since the output is very short (but I don't know if that's what you need)
![20231101_193701_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/10fc6fa2-28ca-42f1-a80f-43690389832b)

**Comment by cagnulein** [comment]
mmm interesting: are you able to pull the .wolflogs folder from the tablet? or at least send me some pictures of the most recent file?

**Comment by glotano** [comment]
Hi @cagnulein, not knowing whether it was a directory or a file (or a directory with similarly named files), I searched for anything with a name similar to *wolflogs in the system. Unfortunately, there is no such thing.
The picture is a little blurry, but anyway I used the command find / -type f,d,l -iname "*wolflogs"
P.S.: I realized in this moment you already write that's a folder :-) anyway I searched for all. That's not the only command I used but it should be enough to realize that the folder, unfortunately, doesn't exist.
![20231101_205729_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/7a42ee34-3ec6-4132-a431-b54e3aad5833)

**Comment by cagnulein** [comment]
the picture shows that find command is not available on your system.

you have to use "ls"

.wolflogs (with the "." at the beginning), it's usually on the root folder

show me your root folder

Il giorno mer 1 nov 2023 alle 21:05 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein, sorry, I didn't realize! Thank you.
Anyway, I searched in the root folder and unfortunately it doesn't exist.
![20231101_211608_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/7df756e6-c183-4b2a-9e8f-0852679a91e4)

**Comment by cagnulein** [comment]
it's an hidden folder, you can't see it with ls without arguments.

did you try to do

ls /. and then press tab multiple times?

Il giorno mer 1 nov 2023 alle 21:23 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
Unfortunately the terminal I installed doesn't have the tab but I tried to write the right name. From the root I typed "ls .wolflogs", the same if type "ls *wolflgogs"
 
![20231101_213134_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/3052c85d-13c6-4dce-993c-63ce0e3a4c8e)
![20231101_214105_HDR_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/28aed8aa-be95-47fe-9b11-677dbc68ba25)

**Comment by cagnulein** [comment]
Can you try to search any log that you can find?

**Comment by glotano** [comment]
Of course @cagnulein, I just tried to find the system logs in /var/log/messages without luck.
Tomorrow I try to understand find more. 
Thank you!

**Comment by glotano** [comment]
Hi @cagnulein, Reading the companion dump logs, I understood that it writes the logs to the /mnt/sdcard directory. I did a manual workout to generate new logs.
The most recent files should be the ones I attached but please let me know if you need the whole folder.
Thanks
Regards
Giovanni
P..S.: Please note that since I didn't find a file named like crashed Crash-2023-11-02, I attached all the file, named like crash*, I found, 
 
[FitPro_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239719/FitPro_02Nov2023.txt)

[Error_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239726/Error_02Nov2023.txt)
[Workout_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239728/Workout_02Nov2023.txt)
[Startup_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239747/Startup_02Nov2023.txt)
[Wifi_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239751/Wifi_02Nov2023.txt)
[Updating_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239773/Updating_02Nov2023.txt)
[Navigation_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239828/Navigation_02Nov2023.txt)
[MvxTrace_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239903/MvxTrace_02Nov2023.txt)
[HeartRate_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239949/HeartRate_02Nov2023.txt)
[FitPro_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240021/FitPro_02Nov2023.txt)
[Fatal_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240089/Fatal_02Nov2023.txt)
[Error_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240117/Error_02Nov2023.txt)
[Dashboard_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240154/Dashboard_02Nov2023.txt)
[Crash-2023-11-01 18.16.21.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240720/Crash-2023-11-01.18.16.21.txt)
[Crash-2023-11-01 18.14.47.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240721/Crash-2023-11-01.18.14.47.txt)
[Crash-2023-11-01 18.14.47 (1).txt](https://github.com/cagnulein/qdomyos-zwift/files/13240722/Crash-2023-11-01.18.14.47.1.txt)
[Connection_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240728/Connection_02Nov2023.txt)
[Combined_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240729/Combined_02Nov2023.txt)
[Ble_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240733/Ble_02Nov2023.txt)

[Benchmark_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13240737/Benchmark_02Nov2023.txt)

**Comment by glotano** [comment]
Hi @cagnulein, I confirm. I also attach two pics to confirm it.
![Screenshot 2023-11-02 153747](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/dc42a267-c42c-42ba-8267-7127af9cb181)
![Screenshot 2023-11-02 153628](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/904a949f-8643-427a-ae3c-54cb924bddce)

**Comment by glotano** [comment]
Hi @cagnulein , we already tried with c1750, but the result is the same. I tried also using s22i, Gran tour pro bike, Proform Studio Bike Pro22, and also using the setting for 2950 treadmill. I tried again to user c1750, but the result is the same (picture attached). If you look to one of the first pics I sent you, (I was using Proform Studio Bike Pro 22 at that time) it always wrote the files in the sdcard folder.
Anyway, only to let you know that the operating system started having problems since I installed the software.
Before it would crash the Ifit app from time to time, now it starts giving problems to the OS itself (as per the attached picture).
The problems disappear every time I reset the machine to factory settings
Thanks
Regards
Giovanni
![20231102_162629_HDR_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/417acb38-c6ac-46ff-ac3a-ff92ea9b3b83)
![20231102_173302_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/e92f85ff-ccea-4ba5-bdbd-fdbb21b06757)

**Comment by cagnulein** [comment]
hi yes i remember but i don't have a picture of the test. in  this picture
i see device: null, are you sure you rebooted the device after selecting
c1750?

looking at the log, the c1750 should be the right one for your device

let me know

Il giorno gio 2 nov 2023 alle 17:50 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
Hi @cagnulein , sorry, I took a picture before rebooting and another after rebooting and at first I had uploaded the one made before the reboot. Now I have uploaded the correct one. Unfortunately, it persists in writing the files to the "sdcard" directory.
If I understand correctly is that the problem?

**Comment by cagnulein** [comment]
ok so now the companion is reading the metrics correctly, tomorrow i will have to check because it still reading again and again the same line.

please keep the c1750 from now on 

thanks
have a nice evening, we are near!

**Comment by glotano** [comment]
Hi @cagnulein, I will keep it set on c1750. please take your time, and let me know in case you need me to do anything.
I wish you a good evening
Thanks to you!
Giovanni

**Comment by cagnulein** [comment]
@glotano checked again the log and i don't undestand why it's not working...the only reason that could explain is that the metrics on the ifit log [FitPro_02Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13239719/FitPro_02Nov2023.txt) are written only when the workout is done.

Could you please try to make a video while you're riding, clicking the dump log several time showing the text box of the companion? the timestamp of the metrics should change

Let me know

**Comment by glotano** [comment]
Hi @cagnulein, 

https://github.com/cagnulein/qdomyos-zwift/assets/149379537/fdd84333-406c-432a-9cdd-67b70823faf2

I know there was no need but I reset the bike to factory settings once again, enabled usb debug mode and installed companion using the apk (via adb no way). 
I started a manual workout. Clicking "Dump Log" several times, it goes back to the beginning of the file, so I had to scroll with my finger moving the video a lot. I reduced the resolution and deleted the audio so the file is much smaller. To me it seems to write during the session. I hope you can see despite me moving a lot.
Thanks again

Giovanni

**Comment by cagnulein** [comment]
it seems that only gears and resistance are moving in realtime, do you get the same reading the log? i mean i'm checking the timestamp of every changed lines in the dump log text box

**Comment by glotano** [comment]
@cagnulein 
The updates values are:
KPH
Grade
CurrentGear
Resistance
I attach a couple of pics from which you could see better
![20231103_154747](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/8dd417da-df01-4b27-8b2b-d1ec182bcdff)
![20231103_154524](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/7ef5a42e-e5cd-42b1-b3da-6e22b27cae0f)

**Comment by glotano** [comment]
@cagnulein, Sorry, I misspoke. The values are there but they are never updated. Now I tried again and scrolled through the whole log. You are right, only gear and resistance are updated!

**Comment by cagnulein** [comment]
but in the log that you sent me attached here I saw the value changing.

So I guess Ifit writes it Only at the end of the workout.

So if I'm right, edit the log file of iFit while you are riding I guess
there are no valid speed for example

Il giorno ven 3 nov 2023 alle 16:15 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein, I did another session, it record all the values.
In my opinion rhe problem is only that Qz app is not able to read the file generated by companion, over UDP
![20231103_162855](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/0a52dc3d-c523-440d-896f-44c4e34392ad)
![20231103_162848_HDR](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/62983622-0ad4-4b17-8ef0-933c1ce49c6c)
![20231103_162936_HDR](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/70e05344-f2c1-4c66-b062-8cdb542d50f0)
![20231103_162900](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/be45da00-4d13-441d-9601-46f7f3d1ca39)

**Comment by cagnulein** [comment]
that's perfect, do you have the qz debug log from this session?

Il giorno ven 3 nov 2023 alle 16:36 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
I don't know because it's really hard to stay at the end of the file scrolling with the finger. The values in the pics could be from the session before. I also did another try. In the recorded session I don't see any trace of KPH. Me too start to thing that the problem is that. In case the problem is that it writes all the values only when I stop the iFit manual workout, in what way we could fix?
P.S.: adding that sometime when I open companion in the tests I found it not responding
![Uploading 20231103_174249.jpg…]()

**Comment by cagnulein** [comment]
ok the workaround should be using logcat directly 

while you are riding try to run logcat and check if you get real time value there

**Comment by cagnulein** [comment]
@glotano check this https://github.com/cagnulein/QZCompanionNordictrackTreadmill/issues/65#issuecomment-1795078043

**Comment by glotano** [comment]
Hi @cagnulein, I don't know how to thank you! I followed that thread so I was able to bind the adb service on port 5555 running:
setprop service.adb.tcp.port 5555;
I uninstalled companion (most probably not needed) so I was able to connect the console's bike from my pc to run
adb shell pm grant org.cagnulein.qzcompanionnordictracktreadmill android.permission.READ_LOGS
(strange but true, I didn't receive any read-only error following this way).
![20231107_123605_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/5f091b18-7829-4ae2-806a-41474db951c1)


Companion started to got the UDP broadcast messages from my phone.
![Screenshot 2023-11-07 131028](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/8e9b785f-4624-423b-aa26-3913f4333fd3)


There is only one problem left.
On qz fitness I don't get the value of watts if I set it to "bike", 
![Screenshot_20231107-123653_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/74a23da0-f2bf-40ed-8e5a-18eff64cd85e)

![Screenshot 2023-11-07 131850](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/7978c78c-d4d6-4515-b8d8-709e453c8325)


while I get the value of watts if I set the app to "treadmill" but of course the value is out of wacky, since it reports to me the power in watts generated by a person who is running on a treadmill while I would like the power expressed in watts of a person who is cycling.
![Screenshot_20231107-124113_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/4109e0f2-bcb3-4c54-a8a3-5f37f35ebede)

![Screenshot_20231107-124121_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/678d8833-d29d-47e0-a6c8-4f83c2017d48)

![Screenshot_20231107-124047_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/0ae69232-c02d-4d21-9577-1734bc856a38)

I will give you an example, a person running at 12 kph on a treadmill using a 5 percent incline will generate really high power, while a person who cycles at the same speed using the same inclination, will generate less than half of watt. 
What happens on Zwift, if I set the qz fitness app to bike, it won't let me cycle since it doesn't detect the watts value. If I set the app on treadmill, it does, but, because Zwift is based on the concept of FTP (Functional Threshold Power) which is the maximum power that a runner is able to express on average and consistently for an hour, as you can guess I always result first in races with a speed that is not the real one and Zwift put me from the first hour of training in the group of people who have the best ftp value.
I tried everything to be able to get the real watts value for a bike, but I just couldn't do it. If we could solve this last problem, I could enjoy Zwift. 
Thank you again!
Regards
Giovanni

**Comment by cagnulein** [comment]
that's great! i will check your past log tomorrow, maybe it's a stupid
issue now

Il giorno mar 7 nov 2023 alle 13:25 glotano ***@***.***> ha
scritto:

**Comment by cagnulein** [comment]
from the https://github.com/cagnulein/qdomyos-zwift/files/13239719/FitPro_02Nov2023.txt I can't see the wattage wrote in the log. SO i guess we have to derive it from the speed. WHat do you think?
It's strange because the others bike do this

**Comment by glotano** [comment]
Hi @cagnulein, I think the same because as you said companion doesn't write any value about the wattage so I think (please correct me if I'm wrong) that's a calculation that the qz fitness app does using the 4 values provided by companion. I think there is a formula. This thought comes from the fact that on companion I always have the c1750 set, changing the treadmill/ bike settings on the app I get/do not get the watts value. That's why I think companion has nothing to do with it. It would be great to be able to get it, I just miss this one thing and finally I can enjoy my bike with Zwift :-)

**Comment by cagnulein** [comment]
the wattage that you see for the treadmill is a calculation.
it would be great if you can send me a table with cadence resistances and
wattage so i can add it to qz in order to calculate the power for your bike

what do you think?

Il giorno mar 7 nov 2023 alle 14:52 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein, thanks for confirming. What you are suggesting makes me really happy. 
To get the values and provide them to you, should I do a session with iFit and find them somewhere in the app installation files on the Android system? Please just tell me what I need to do and I will do it. That would make me really happy. Thank you! You are great

**Comment by glotano** [comment]
@cagnulein, I was thinking the following.
Since, as you said, for the other versions of my bike qz app already calculates the wattage, couldn't you apply the same formula for mine? They differ only in monitor size, otherwise they are almost identical (at least for the features we are interested in), they differ in some details that we don't care about, type of pedals and whatnot.
It's important to have taken the inclination (other than the cadence and resistance) into account when calculating the watts delivered, as I think it is the one bike that has it (maybe Peloton as well but not others).
Please let me know.

**Comment by cagnulein** [comment]
nope, the other bike version are writing wattage in the log directly,
that's the main difference  ;)

Il giorno mar 7 nov 2023 alle 23:30 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein I tell you this since the argument is quite simple.
It's well established that the calculation of watts must be done through a formula since companion does not provide that value.
That value is not written for any Nordictrack Studio Bike. It should be a calculation built into the app qz fitness, which in my opinion has not been done. It's evident from the fact that having the same values as input, 
KPH
Grade
CurrentGear
Resistance
if you set the app to "treadmill" the power value is returned (but for a runner) while if you set it to bike, it's not returned. Missing in the app is the proper formula to return the actual value expressed in Watts for a Bike.
This problem affects all the Nordictrack Studio bikes.
I think this is something you could implement to let the Nordictrack bikes to be really compliant with Zwift, providing to the app the real wattage for a cyclist and not for a runner.

**Comment by cagnulein** [comment]
> @cagnulein, also the mine writes the wattage in iFit. Which setting have set on companion the owners of the s22 to write the wattage? 

Nothing particular, they wrote it by default! That's the issue :(

> The bike is exactly the same, only the monitor size changes. I have the feeling the owners of Nordictrack studio bikes, are using the bike as a threadmill generating the wattage of a runner.

nope I can tell you for sure that's not the case. I have the log of the others bike. Maybe changing the workout type in ifit you can have the wattage on the log? You can try to experiment on this on your side


> @cagnulein I tell you this since the argument is quite simple. It's well established that the calculation of watts must be done through a formula since companion does not provide that value. 

yes i'm alrady doing this for other bikes and treadmills, but every bike is different so I need a table in order to add it if you can have wattage in the log.

> This problem affects all the Nordictrack Studio bikes.

Nope, only yours as far as I know

> think this is something you could implement to let the Nordictrack bikes to be really compliant with Zwift, providing to the app the real wattage for a cyclist and not for a runner. Please, if you will decide to implement the formula for Nordictrack bikes, don't forget that you should take in consideration also the inclination since that's probably the only bike that owns that feature and that obviously affects along with cadence, and resistance ont the power expressed in watt.

I can add it easily but I need the table from you!

**Comment by cagnulein** [comment]
> @cagnulein, also the mine writes the wattage in iFit. Which setting have set on companion the owners of the s22 to write the wattage? 

Nothing particular, they wrote it by default! That's the issue :(

> The bike is exactly the same, only the monitor size changes. I have the feeling the owners of Nordictrack studio bikes, are using the bike as a threadmill generating the wattage of a runner.

nope I can tell you for sure that's not the case. I have the log of the others bike. Maybe changing the workout type in ifit you can have the wattage on the log? You can try to experiment on this on your side


> @cagnulein I tell you this since the argument is quite simple. It's well established that the calculation of watts must be done through a formula since companion does not provide that value. 

yes i'm alrady doing this for other bikes and treadmills, but every bike is different so I need a table in order to add it if you can have wattage in the log.

> This problem affects all the Nordictrack Studio bikes.

Nope, only yours as far as I know

> think this is something you could implement to let the Nordictrack bikes to be really compliant with Zwift, providing to the app the real wattage for a cyclist and not for a runner. Please, if you will decide to implement the formula for Nordictrack bikes, don't forget that you should take in consideration also the inclination since that's probably the only bike that owns that feature and that obviously affects along with cadence, and resistance ont the power expressed in watt.

I can add it easily but I need the table from you!

**Comment by glotano** [comment]
@cagnulein, do you know where I can find the table?

**Comment by cagnulein** [comment]
> @cagnulein, do you know where I can find the table?

you have to record it :)

a table with

cadence resistance level and power, something like this https://github.com/cagnulein/qdomyos-zwift/issues/1683#issuecomment-1754163437

**Comment by glotano** [comment]
@cagnulein cagnulein, I have alredy some from my previous workouts but I will try to do a small session to provide you with data that can help you better than what I have now. I'm going to have a session with all the resistance and inclinations, trying to be as regular as possible about the cadence, in any case the data that will come out will be very detailed. I hope to have some time to do it this evening.

**Comment by cagnulein** [comment]
ok no rush from me, try also to collect more cadence for the same resistance, so we will have a more detailed table.
Thanks

**Comment by glotano** [comment]
@cagnulein, I did a one-hour workout. The first 20 minutes I devoted exclusively to collecting the data I had to provide, and after that I alternated between the real workout and the data collection.
I used all possible incline levels as well as resistance.
I am passing you the raw file since putrously, opening it with excel I always have the same timestamp, while opening it through notepad (which for obvious reasons is inconvenient) shows that it is exactly the file related to the data collection that lasted an hour.
The start time stamp, from notepad is 15:58:56.358 and the end time stamp is 16:59:04.358.
You probably have a better way to open the file. 
I didn't remove any columns, so you have everything but the main ones are:
ns1:Watts
ns1:Resistance
ns1:Cadence
probably ns1:AltitudeMeters (my expectation was to found incline, instead)
![databike](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/117a1d8c-1758-4799-8130-45029349fffb)


The thing that I was so sorry about is that I couldn't find the column for inclination, the spin bike has tilt levels ranging from -10 to 22 which affect the watts output so much. I realized that instead they included the gain/loss altrimetric. 
I hope you can do something to include the tilt value, otherwise it would become a useless job. That is why I am also sending you the file generated by companion that should include that value.
P.S.: please rename the extension file of the named 2023_11_08_16_11_Manual_Workout into tcx. I needed to add a fake extension (.log) to send it. Github doesn't allow to download the file with that extension


[2023_11_08_16_11_Manual_Workout.log](https://github.com/cagnulein/qdomyos-zwift/files/13301055/2023_11_08_16_11_Manual_Workout.log)
[FitPro_08Nov2023.txt](https://github.com/cagnulein/qdomyos-zwift/files/13301083/FitPro_08Nov2023.txt)

**Comment by cagnulein** [comment]
Thanks, I'll check it out tomorrow

Il giorno mer 8 nov 2023 alle 21:31 glotano ***@***.***> ha
scritto:

**Comment by glotano** [comment]
@cagnulein, thanks to you! Have a goodnight.

**Comment by cagnulein** [comment]
thanks @glotano the data seem fine, could you summarize them in a table as I posted previously?
I think you can use chatgpt4 to create it for example or of course manually 
Let me know

**Comment by cagnulein** [comment]
@glotano are you still interesting on this?

**Comment by glotano** [comment]
@cagnulein, the problem was not related to importing the data in a table. It can be done with Excel. I'm not a developer but the main problem in my opinion is how the log express the inclination As said with this kind of bike there is to take into the account also the inclination (from -10 to +20). It affects the expressed wattage a lot, with the other two values (cadence,resistance).
If you want to give it a try I attach the file, already imported into a table. I have excluded the irrelevant values. The main problem, in my opinion is how to create the formula taking into account the inclination. The fact that it is expressed in an altitude gain or a decrease, I think makes things more difficult for you. But, if you can do it, I would be glad.
Thank you!
[IfitLog.xlsx](https://github.com/cagnulein/qdomyos-zwift/files/13467815/IfitLog.xlsx)

**Comment by cagnulein** [comment]
as far as i know, on your bike, when the inclination changes, it changes also the resistance, isn't it? if so we can just put off the game the inclination, and use the resistance and cadence only. Don't you think?

**Comment by glotano** [comment]
@cagnulein , If you only take into account cadence and resistance, you will not have the real power expressed in watts.
My bike works the following way:
by changing the inclination you have a change in the resistance felt on your legs (the physical effort) but not a change in the resistance value (since it's an input you decide, not an output like the wattage) and everything is reported, accordingly, in watts, because iFit takes into account cadence, resistance, inclination.

I give an example, if I set resistance at 5 and cycle with a cadence of 30, I will generate a power of x watts. By increasing the incline or decreasing it, while maintaining the same cadence and resistance, the power in watts expressed, will increase of x+y or decrease x-y . 
If you don't take into account the inclination, in the formula, the calculation of watts will not be the real one at all.
What will happen is, I cycle with a cadence of 30 and with a resistance of 5 generating, say 100 watts, increasing the inclination to 10, I will generate 200 watts (as iFit does) but if you don't take into account the inclination, qzApp will always give me 100.
No, that would not work at all.

**Comment by cagnulein** [comment]
ok that's strange. It means that it worked in a different way from all the other nordictrack bikes. But ok. So we have to find a way to read the power somehow.
Can you experiment with adb if you can have that value? It also strange that this bike doesn't write this in the log since all the other one does. Maybe it's system language that gives a different behavior (italian instead of english? i'm just guessing)

**Comment by glotano** [comment]
@cagnulein, during my workout I took two pics to sommerize what I explained above. Please look at the wattage, same cadence, same resistance, with two different inclination level

![20231127_180517_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/5d5a56e7-a8a4-4126-957a-3238bd452592)
![20231127_180444_resized](https://github.com/cagnulein/qdomyos-zwift/assets/149379537/85f6d4af-9814-4a33-9e9a-93878295e008)
Unfortunately the value of the inclination is nowhere. The iFit software provide the altitude gain instead of the inclination level.

**Comment by glotano** [comment]
@cagnulein. It would be a useless effort. We will keep each other updated in case one or the other has any ideas to get an accurate calculation of power output.
Thanks
Regards
Giovanni
